feat: log in with ntnui medlem

eivindkopperud opened the pull request:
Closes #21 
## Summary of changes
Extends the Form component with input fields for phone number and password. When successful login, the name, email and bank account number gets filled in from the user's account.

React masters @janewl and @LiviaValenti - can you take a look at this? The only React experience I have is 1week in a summer internship so I don't know anything:))<3

## Images
<details>
 <summary> Upon entering the app </summary>

![image](https://user-images.githubusercontent.com/43027083/161422524-f06076e0-06d1-4e10-a275-9047da8426c3.png)
</details>

<details>
 <summary> Upon clicking "Hent data fra medlem.ntnui.no </summary>

![image](https://user-images.githubusercontent.com/43027083/161422556-ef5e706e-dee0-4349-b369-ce10c11ad39d.png)
</details>



## Todo
- [x] Fetch the user's group and change the group input to a dropdown with all the group values
- [x] Auto fill the cashier email based on the group
- [x] Maaaybe refactor the login to an own component?
- [x] Make the requests prettier, maybe extracting to a requiests.ts file
- [x] Clean up my code as it is trash atm

eivindkopperud committed:
feat: log in with NTNUI medlem

LiviaValenti commented on webapp/components/Form.tsx:
```diff
@@ -65,6 +110,26 @@ const Form = (): JSX.Element => {
+      helperText="Ditt passord på medlem.ntnui.no"
+      type="password"/>
+      <Button variant="contained"
+        color="primary"
```
Trivial tip: Since we use mui and we have the icon-pack, could you perhaps add an icon on the button? E.g. "FileDownload" 🤩

```suggestion
        color="primary"
        startIcon={<FileDownload/>}
```

LiviaValenti commented on webapp/components/Form.tsx:
```diff
@@ -57,6 +66,42 @@ const Form = (): JSX.Element => {
     </div>
   );
 
```
I agree on refactoring the login to a separate component. It may not be that important e.g. if you decide to extract the API-logic to a services-file or such. Doing both is good though! (refactoring and extracting logic) ⭐  
 
> Maaaybe refactor the login to an own component?
> Make the requests prettier, maybe extracting to a requiests.ts file

LiviaValenti commented on webapp/components/Form.tsx:
```diff
@@ -65,6 +110,26 @@ const Form = (): JSX.Element => {
+      type="password"/>
+      <Button variant="contained"
+        color="primary"
+        disabled={submitting || success == true}
```
You need to toggle the `submitting` and `success`-attributes in your fetchUserData if you want it to be functional. (Take a look at how it is done on line 211). The API-logic lying inside the button-component was there before this PR, but it really should be refactored into a method (as you have done with fetchUserData) ✂️ 

At the moment: If I click the button to retrieve data and it is succesfull it is not disabled and I can continue to send requests. It would also be nice to display an error-message on error too. If the request fails now I get the following:

![image](https://user-images.githubusercontent.com/54808860/161424303-8d7f0e3d-25f0-4aa9-8f04-d6197dc92ae6.png)

LiviaValenti commented on webapp/components/Form.tsx:
```diff
@@ -57,6 +66,42 @@ const Form = (): JSX.Element => {
+      setAccountNumber(json.bank_account_number)
+
+    }
+    catch (err) {
```
From [fetch-docs](https://developer.mozilla.org/en-US/docs/Web/API/fetch)

> A [fetch()](https://developer.mozilla.org/en-US/docs/Web/API/fetch) promise only rejects when a network error is encountered (which is usually when there's a permissions issue or similar). A [fetch()](https://developer.mozilla.org/en-US/docs/Web/API/fetch) promise does not reject on HTTP errors (404, etc.). Instead, a then() handler must check the [Response.ok](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok) and/or [Response.status](https://developer.mozilla.org/en-US/docs/Web/API/Response/status) properties.


Thus triggering 401 will not go to the catch-clause and you set the name, mail and account number even if there is none to set.

Add a check if the response-object is ok or not like they have done further down in this file or as explained in [the fech-docs](https://developer.mozilla.org/en-US/docs/Web/API/fetch#examples).

eivindkopperud replied:
> If you are to do this, you need to do some good error-handling if something fails on the retrieval of groups. If the user decides _not_ to use the login-to-retrieve-data-functionality. You must display all groups. [Mui has some nifty autocomplete-components ](https://mui.com/components/autocomplete/) which would make retrieving all groups and searching in them fast anyway. Retrieving only the groups you are member of is a nice feature, though ⭐

Would you consider it a bad practice to change the component to a MUI Dropdown if the user has memberships in one group or more? E.g. something like this:

```javascript
groups = fetchGroups()
{ groups.length >= 1 ? <Dropdown prop1=".." prop2=".." /> : <Input prop1=".." prop2=".."/>
```
Or maybe it is just simpler to fetch all groups either way and use the same dropdown whether the user fetches data or not. 

> I see nothing wrong with this one other that it will require one request every time you change the group-dropdown which again requires some proper error-handling. Nice feature ⭐

What if we have a `[groups, setGroups] = useState([])` which contains all the groups? Here we will also have the `gsuite_prefix` which we can use for the cashier email, i.e. 
`setMailto('${selectedGroup.gsuite_prefix}-kasserer@ntnui.no')`

LiviaValenti replied:
> change the component to a MUI Dropdown if the user has memberships in one group or more

If the user has memberships in no groups it would be a normal input-field, not a dropdown? Perhaps it would be better to just give the user a dropdown of all groups in that case? Or am I misunderstanding?

It might better to just fetch all groups. It will be an improvement from before either way.

> What if we have a [groups, setGroups] = useState([]) which contains all the groups?

That would be the easiest solution I believe 🧷

eivindkopperud replied:
Nice! I think we're onto something good 🤓 Will fix 💪

LiviaValenti reviewed (commented):
The happy-path works well and I believe the users will be suuuuper happy for this implementation. ✨ 

Here are some comments on the To Dos that were not adressed in other comments:

>  Fetch the user's group and change the group input to a dropdown with all the group values

If you are to do this, you need to do some good error-handling if something fails on the retrieval of groups. If the user decides *not* to use the login-to-retrieve-data-functionality. You must display all groups. [Mui has some nifty autocomplete-components ](https://mui.com/components/autocomplete/) which would make retrieving all groups and searching in them fast anyway. Retrieving only the groups you are member of is a nice feature, though ⭐ 

> Auto fill the cashier email based on the group

I see nothing wrong with this one other that it will require one request every time you change the group-dropdown which again requires some proper error-handling. Nice feature ⭐

Xtrah commented:
I think this is a GREAT feature for this app that will be very appreciated by both users and cashiers 🤩

I wonder though if the current design makes it look like you are required to login to use the form? Can we hide the login fields in a dropdown or popup, or somehow make it clearer that its a quality of life feature and optional to use? 🤔

eivindkopperud committed:
just fix everything (:

eivindkopperud commented:
#⚠️ THOSE WHO ARE ON VACATION (e.g. in Italy 🍝) ARE NOT ALLOWED TO REVIEW ⚠️ 

I've not added a new suggestion. Some things I've fixed are:

## An accordion 
When the user enters, the accordion is not expanded. The user can click on it to extend it, see the picture below.
![image](https://user-images.githubusercontent.com/43027083/162575420-55bf49c1-8f7a-4902-a386-1b95ef16740b.png)

If the user information is incorrect, an alert is shown:
![image](https://user-images.githubusercontent.com/43027083/162575469-62adb247-e996-4988-a598-43194b857f25.png)

If successful, the accordion is rolled up and disabled. **_Although I'm not sure if the disabling is actually wanted?_**
The user data (and bank account number, if existing) is then filled in the forms.

This accordion is extracted to it's own component, called `LogInAccordion`

## Group dropdown
Fetches all groups. I did not fetch _only_ the user's groups if they logged in, because you miiiight want to refund for some other groups. And you might not be in the group in the membership system. So I just have a searchable dropdown with all groups.

If no groups are available, the regular input field is shown, and you can type freely.
The cashier mail is also updated based on the group. It is a bit whacky now maybe since it a group change changes the value in the input field.

## medlem.ntnui.no request extraction
I've also extracted all request functions to a separate file. We might want to use NTNUI/ntnui-tools for it later 🚀

eivindkopperud committed:
fix change group bug

matsfinsas commented on webapp/components/GroupDropDown.tsx:
```diff
@@ -0,0 +1,53 @@
+    name="Gruppe"
+    required={true}
+    updateForm={updateForm}
+    helperText="Som utgiften skal betales avhit"
```
```suggestion
    helperText="Som utgiften skal betales av"
```

matsfinsas commented on webapp/components/GroupDropDown.tsx:
```diff
@@ -0,0 +1,53 @@
+            renderInput={(params) => <TextField {...params} label="Gruppe" variant="outlined"/>}
+        />
+        <FormHelperText style={{marginLeft: '1em'}}>Som utgiften skal betales av</FormHelperText>
+    </Box>
```
Nice to get a list of groups, but maybe allow custom input? Now it blanks out if it doesn't match a group in the list.

eivindkopperud replied:
Fixed by @janewl amazing Material UI documentation reading skills 📈

matsfinsas commented on webapp/components/Form.tsx:
```diff
@@ -46,6 +53,18 @@ const Form = (): JSX.Element => {
+
+  const updateGroup = (group: string) => {
+    setGroup(group)
+    setMailto(`${group.toLowerCase()}-kasserer@ntnui.no`)
```
This doesn't work correctly for group names with spaces, such as Ski og Fjellsport would return `ski og fjellsport-kasserer@ntnui.no`, and gsuite_prefix might be customized!

eivindkopperud replied:
I'll leave the gsuite prefix stuff for later as I'm not sure how I pass the group reference around in the <Autocomplete> component. See #26

matsfinsas commented on webapp/components/LogInAccordion.tsx:
```diff
@@ -0,0 +1,89 @@
+            style={{ width: '100%', marginTop: '1em', marginBottom: '1em' }}
+            className={styles.fullWidth}
+            onClick={() => fetchUserData()}
+        >
```
When writing username and password, clicking enter should click this button ⌨️

eivindkopperud replied:
> Building the Docker image also gives an error after your changes 🤔
> 
> ```
> #23 2.103 info  - Checking validity of types...
> #23 9.780 Failed to compile.
> #23 9.780
> #23 9.781 ./pages/_document.tsx:4:26
> #23 9.781 Type error: Type '() => any' is not a constructor function type.
> #23 9.781
> #23 9.781   2 | import Document, { Html, Head, Main, NextScript } from 'next/document';
> #23 9.781   3 |
> #23 9.781 > 4 | class MyDocument extends Document {
> #23 9.781     |                          ^
> #23 9.781   5 |   // eslint-disable-next-line
> #23 9.781   6 |   static async getInitialProps(ctx: any) {
> #23 9.781   7 |     const initialProps = await Document.getInitialProps(ctx);
> #23 9.829 error Command failed with exit code 1.
> ```

I think this is due to the wrong next version, see #22

matsfinsas commented on webapp/components/LogInAccordion.tsx:
```diff
@@ -0,0 +1,89 @@
+        <AccordionSummary
+            expandIcon={<ExpandMoreIcon />}
+            >
+            <Typography>Auto-fyll med data fra medlem.ntnui.no (valgfritt)</Typography>
```
```suggestion
            <Typography>Hent informasjon fra medlem.ntnui.no (valgfritt)</Typography>
```
Maybe?

matsfinsas commented on webapp/components/LogInAccordion.tsx:
```diff
@@ -0,0 +1,89 @@
+import React, { useState } from 'react';
+import { Typography, Button, Accordion } from '@material-ui/core';
+import AccordionSummary from '@material-ui/core/AccordionSummary';
+import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
+import Alert from '@material-ui/lab/Alert';
+
+
+
+import GetAppIcon from '@material-ui/icons/GetApp';
+
+
+import Input from './Input';
+
+import styles from './Form.module.css';
+import { fetchProfile, fetchToken } from 'service/medlemService';
```
We should probably add some code formatting config/tool 😁

eivindkopperud replied:
Hehe, agree. Will create an issue, but attempt to clean up some stuff manually 😂 
See #25

matsfinsas reviewed (commented):
✨ Frontend developer ✨

Xtrah commented:
Building the image using the also gives an error after your changes 🤔

```
#23 2.103 info  - Checking validity of types...
#23 9.780 Failed to compile.
#23 9.780
#23 9.781 ./pages/_document.tsx:4:26
#23 9.781 Type error: Type '() => any' is not a constructor function type.
#23 9.781
#23 9.781   2 | import Document, { Html, Head, Main, NextScript } from 'next/document';
#23 9.781   3 |
#23 9.781 > 4 | class MyDocument extends Document {
#23 9.781     |                          ^
#23 9.781   5 |   // eslint-disable-next-line
#23 9.781   6 |   static async getInitialProps(ctx: any) {
#23 9.781   7 |     const initialProps = await Document.getInitialProps(ctx);
#23 9.829 error Command failed with exit code 1.
```

eivindkopperud committed:
style loginaccordion

eivindkopperud committed:
enhance wonky group dropdown clearing

eivindkopperud committed:
Update webapp/components/GroupDropDown.tsx

Co-authored-by: Mats Finsås <xtrah@live.com>

eivindkopperud committed:
add margin to alert component

eivindkopperud committed:
fix space in phone number bug

eivindkopperud committed:
slugify group in cashier email

eivindkopperud committed:
manual formatting

eivindkopperud commented:
> Building the Docker image also gives an error after your changes 🤔
> 
> ```
> #23 2.103 info  - Checking validity of types...
> #23 9.780 Failed to compile.
> #23 9.780
> #23 9.781 ./pages/_document.tsx:4:26
> #23 9.781 Type error: Type '() => any' is not a constructor function type.
> #23 9.781
> #23 9.781   2 | import Document, { Html, Head, Main, NextScript } from 'next/document';
> #23 9.781   3 |
> #23 9.781 > 4 | class MyDocument extends Document {
> #23 9.781     |                          ^
> #23 9.781   5 |   // eslint-disable-next-line
> #23 9.781   6 |   static async getInitialProps(ctx: any) {
> #23 9.781   7 |     const initialProps = await Document.getInitialProps(ctx);
> #23 9.829 error Command failed with exit code 1.
> ```

I think this is due to the wrong next version, see #22

eivindkopperud commented:
I've fixed some stuff now, do you mind taking a final look? I'd love to get some comments about all the best practices I don't follow here as well @LiviaValenti @janewl 😉

Mortefal commented on webapp/service/medlemService.ts:
```diff
@@ -0,0 +1,56 @@
+const API_URL = 'http://localhost:8000'
```
localhost?

eivindkopperud replied:
Thanks for the reminder, I'll switch it to api.ntnui.no before merging 😁

Mortefal commented on webapp/service/medlemService.ts:
```diff
@@ -0,0 +1,56 @@
+    return await response.json();
+}
+
+export const updateProfile = async (token: string, accountNumber: string) => {
```
https://www.npmjs.com/package/iban used this package when working with bank, it works great :)

eivindkopperud replied:
Nice! Referenced this in a new issue at #27

eivindkopperud committed:
update api url to api.ntnui.no

Commit:
Merge branch 'refusjon-master' into feat/refusjon-log-in

matsfinsas commented on webapp/service/medlemService.ts:
```diff
@@ -0,0 +1,56 @@
+const API_URL = 'https://api.ntnui.no'
```
This could be an optional environment variable maybe? It is built into React, so I think you can just
```suggestion
const API_URL = process.env.REACT_APP_API_URL || 'https://api.ntnui.no'
```

And add it to the table in the README if you do, together with other relevant info for further development and testing, such as dealing with CORS 😛

matsfinsas commented on webapp/components/LogInAccordion.tsx:
```diff
@@ -0,0 +1,83 @@
+        defaultExpanded={false}
+        disabled={isSuccess}
+        expanded={expanded}
+        style={{marginBottom: '2em', paddingLeft: '0.5em', paddingRight: '0.5em', backgroundColor: '#a1e6c0', borderRadius: '5px'}}
```
Maybe a less dominant background color? 😇 

```suggestion
        style={{marginBottom: '2em', paddingLeft: '0.5em', paddingRight: '0.5em', backgroundColor: '#a1e6c07a', borderRadius: '5px'}}
```

matsfinsas reviewed (approved):
I'm lovin' it Ⓜ️

matsfinsas committed:
Update webapp/components/LogInAccordion.tsx

matsfinsas committed:
Update webapp/service/medlemService.ts

eivindkopperud commented:
@Xtrah Is this good now? 😇 :party_parrot_

Xtrah commented:
I think it looks OK now 🏅
